kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/http/HttpResponse.java

```diff
@@ -63,29 +64,71 @@ default String message() {
    * @throws IOException in case there's an I/O problem
    */
   default String bodyString() throws IOException {
-    Object body = body();
-    if (body == null) {
+    final Object body = body();
+    final SupportedResponses supportedResponse = SupportedResponses.from(body);
+    if (supportedResponse == null) {
       return "";
     }
-    if (body instanceof String) {
-      return (String) body;
-    }
-    if (body instanceof Reader) {
-      return IOHelpers.readFully((Reader) body);
-    }
-    if (body instanceof byte[]) {
-      return new String((byte[]) body, UTF_8);
-    }
-    return IOHelpers.readFully((InputStream) body, UTF_8);
+    return supportedResponse.asString(body);
   }
 
   /**
    * The original {@link HttpRequest} that initiated this response.
-   * 
+   *
    * @return the HTTP request.
    */
   HttpRequest request();
 
   Optional<HttpResponse<?>> previousResponse();
 
+  enum SupportedResponses {
```

Comment:
I did want to expose the SupportedResponses, because besides Javadoc, there's nothing tangible that suggests users what the preferred response should be.
However, a sendAsync method signature accepting the enum would be less friendly since we'd lose the T parameter.

Would making the SupportedResponses.asString and SupportedResponses.sendAsync methods private be enough for you?

Reply:
> I did want to expose the SupportedResponses, because besides Javadoc, there's nothing tangible that suggests users what the preferred response should be.

Do you envision refactoring the calling code from:

client.sendAsync()

To:

SupportedResponses.TEXT.sendAsync(...)

And eliminating HttpClient.sendAsync altogether?  I'm fine with that too.  There will eventually be a couple of other cases - ReadableByteChannel - which could either be blocking (essentially what is underpinning the InputStream now) or non-blocking.

Reply:
well, I was thinking more something like this:
```
  default <T> CompletableFuture<HttpResponse<T>> sendAsync(HttpRequest request, HttpResponse.SupportedResponses type) {
    return type.sendAsync(request, this);
  }
```

I added the method to this PR so you can check

Reply:
Oh, I thought you were saying you didn't want to loose the type parameter from the enum.  Yes, that is fine.  But are you thinking of moving forward with this change in this pr, or a follow-on?  Some tests won't compile with this change in.